In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


# imported data

In [2]:
df=pd.read_csv('Fraud.csv')


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


df['nameOrig'] = df['nameOrig'].astype(str)
df['nameDest'] = df['nameDest'].astype(str)

In [4]:
df.corr()["isFraud"]

C:\Users\sahil pandey\AppData\Local\Temp\ipykernel_636\1163818509.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()["isFraud"]


step              0.031578
amount            0.076688
oldbalanceOrg     0.010154
newbalanceOrig   -0.008148
oldbalanceDest   -0.005885
newbalanceDest    0.000535
isFraud           1.000000
isFlaggedFraud    0.044109
Name: isFraud, dtype: float64

In [5]:
df['type'].value_counts()

CASH_OUT    2237500
PAYMENT     2151495
CASH_IN     1399284
TRANSFER     532909
DEBIT         41432
Name: type, dtype: int64

# Analyzed the data 
Also used pandas profiling for analysing in depth .

In [6]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(df.drop(columns=['isFraud','nameOrig','nameDest']),df['isFraud'],
                                                test_size=0.2)

In [7]:
from sklearn.preprocessing import OneHotEncoder 

In [8]:
from sklearn.compose import ColumnTransformer

In [9]:
transformer = ColumnTransformer(transformers=[  ('tnf3',OneHotEncoder(sparse=False,drop='first'),['type'])
],remainder='passthrough')
    

In [10]:
X_train=transformer.fit_transform(X_train)


C:\Users\sahil pandey\anaconda3\Lib\site-packages\sklearn\preprocessing\_encoders.py:972: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [11]:
X_test=transformer.transform(X_test)

In [12]:
X_train.shape

(5090096, 11)

In [13]:
X_test.shape

(1272524, 11)

# Model training and checking accuracy 

In [14]:
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression

In [15]:
clf = LogisticRegression(random_state=0).fit(X_train,y_train)

In [16]:
y_pred = clf.predict(X_test)

In [17]:
clf.score(X_train,y_train)

0.9982338250594881

In [18]:
from sklearn.metrics import accuracy_score

In [19]:
accuracy_score(y_test,y_pred)

0.998258579013048

This is the model which will predict the fraud transaction with 99.83% accuracy .

I did pandas profiling after analysing the collinearity and other factors , I did one hot encoding for column 'type'.

I used Logistic regression model in this project for best performance .

Key factor was amount and change in bank balance between old and new .

Yes these factors make sense  because ,fraudulent was cashing out the money from diverse type of accounts so corelation helped alot . 

Big transactions ,too small transactions after which new balance becomes zero can be suspecious.

This model is very efficient after it ,for comfirmation we can do KYC of client for 100% confirmation or OTP confirmation . 